### PR TITLE
 service metrics to track availability and latency

### PR DESCRIPTION
**Pull request by venkat1109**

### CODE REVIEW

File: common/metrics/defs.go [hunk 1]
```diff
@@ -200,22 +200,42 @@ const (
...
-	// RespondDecisionTaskCompletedScope tracks RespondDecisionTaskCompleted API calls received by service
-	RespondDecisionTaskCompletedScope
-	// RespondActivityTaskCompletedScope tracks RespondActivityTaskCompleted API calls received by service
-	RespondActivityTaskCompletedScope
-	// RespondActivityTaskFailedScope tracks RespondActivityTaskFailed API calls received by service
-	RespondActivityTaskFailedScope
-	// GetWorkflowExecutionHistoryScope tracks GetWorkflowExecutionHistory API calls received by service
-	GetWorkflowExecutionHistoryScope
+	// FrontendStartWorkflowExecutionScope is the metric scope for frontend.StartWorkflowExecution
+	FrontendStartWorkflowExecutionScope = iota + NumCommonScopes
```
Comment (madhuravi) [inline common/metrics/defs.go]: where are these used?

Comment (venkat1109) [inline common/metrics/defs.go]: see frontend/handler.go:548

File: service/matching/handler.go [hunk 2]
```diff
@@ -74,37 +78,82 @@ func (h *Handler) IsHealthy(ctx thrift.Context) (bool, error) {
...
+		return err
+	case *gen.WorkflowExecutionAlreadyStartedError:
+		h.metricsClient.IncCounter(scope, metrics.CadenceErrExecutionAlreadyStartedCounter)
+		return err
+	case *gen.DomainAlreadyExistsError:
+		h.metricsClient.IncCounter(scope, metrics.CadenceErrDomainAlreadyExistsCounter)
+		return err
+	default:
+		h.metricsClient.IncCounter(scope, metrics.CadenceFailures)
+		return &gen.InternalServiceError{Message: err.Error()}
```
Comment (madhuravi) [inline service/matching/handler.go]: why is this necessary?

Comment (venkat1109) [inline service/matching/handler.go]: some of the layers down the stack return the error as is and they are bubbled all the way up. This is the uppermost layer and we want to make sure we stick to the public exception contract i.e. return one of {BadRequest, InternalService, EntityNotExist}Error etc. Any other errors other than ones listed in the select{} clause, are by, definition, retryable server side errors (i.e. internalServiceErrors)

Comment (madhuravi) [inline service/matching/handler.go]: got it, thank you.
